### PR TITLE
Create new task to remove deploy groups from stages given an owner

### PR DESCRIPTION
**Pull request by lara**
@zendesk/compute 

- Add confirmation step that allows user to proceed with proposed actions
- Optionally delete emptied stages

### CODE REVIEW

File: lib/tasks/maintenance.rake [hunk 1]
```diff
@@ -43,4 +43,38 @@ namespace :maintenance do
       from_stage.destroy! if delete_empty && from_stage.deploy_groups.empty?
     end
   end
+
+  # FROM and TO handles multiple deploy groups to not mess with projects that have each deploy group in a single stage.
```
Comment (grosser) [inline lib/tasks/maintenance.rake:47]: wut ?
... we want to mess with these too

File: lib/tasks/maintenance.rake [hunk 2]
```diff
@@ -43,4 +43,38 @@ namespace :maintenance do
...
+  desc "delete deploy groups across stages for all projects. OWNER= DELETE=dg1,dg2 [DELETE_EMPTY=true]"
```
Comment (grosser) [inline lib/tasks/maintenance.rake:48]: we usually remove just 1 ... and even if we remove 2 I'd rather run it twice
```suggestion
  desc "delete deploy groups across stages for all projects. OWNER= DELETE=dg1 [DELETE_EMPTY=true]"
```

File: lib/tasks/maintenance.rake [hunk 3]
```diff
@@ -43,4 +43,38 @@ namespace :maintenance do
...
+  task delete_deploy_group: :environment do
+    projects = Project.where(owner: ENV.fetch("OWNER"))
+    delete = ENV.fetch("DELETE").split(",")
+    delete_group = DeployGroup.find_by_permalink!(delete)
```
Comment (grosser) [inline lib/tasks/maintenance.rake:52]: find by array does not work
```suggestion
    delete = ENV.fetch("DELETE")
    delete_group = DeployGroup.find_by_permalink!(delete)
```

File: lib/tasks/maintenance.rake [hunk 4]
```diff
@@ -43,4 +43,38 @@ namespace :maintenance do
...
+    delete_empty = (ENV["DELETE_EMPTY"] == "true")
+
+    puts "found #{projects.size} projects"
+
+    actions = projects.map do |project|
+      from_stage = project.stages.detect { |stage| delete & stage.deploy_groups.map(&:permalink) == delete }
```
Comment (grosser) [inline lib/tasks/maintenance.rake:58]:
```suggestion
      from_stage = project.stages.detect { |stage| stage.deploy_groups.map(&:permalink).include? delete }
```

File: lib/tasks/maintenance.rake [hunk 5]
```diff
@@ -43,4 +43,38 @@ namespace :maintenance do
...
+      [project, from_stage]
+    end
+
+    actions.select! do |project, from_stage|
+      if !from_stage
+        puts "#{project.name} Unable to find DELETE"
+      else
+        empty = (from_stage.deploy_groups == [delete_group])
+        puts "#{project.name} to be removed from:#{from_stage&.name}#{" (+ delete stage)" if empty}}"
```
Comment (grosser) [inline lib/tasks/maintenance.rake:67]:
```suggestion
        puts "#{project.name} to be removed from:#{from_stage&.name}#{" (+ delete stage)" if empty && delete_empty}}"
```

File: lib/tasks/maintenance.rake [hunk 6]
```diff
@@ -43,4 +43,38 @@ namespace :maintenance do
...
+    puts "found #{projects.size} projects"
+
+    actions = projects.map do |project|
+      from_stage = project.stages.detect { |stage| delete & stage.deploy_groups.map(&:permalink) == delete }
+      [project, from_stage]
+    end
+
+    actions.select! do |project, from_stage|
+      if !from_stage
+        puts "#{project.name} Unable to find DELETE"
```
Comment (uthark) [inline lib/tasks/maintenance.rake:64]:
```suggestion
        puts "#{project.name} Unable to find DELETE"
        false 
```

make it explicit?

File: lib/tasks/maintenance.rake (hunk 5 above)
Comment (grosser) [reply lib/tasks/maintenance.rake:67]: ... doing the logic twice to decide if we should delete is kinda meh anyway, better to add a true/false flag to the `actions` ?

File: lib/tasks/maintenance.rake (hunk 3 above)
Comment (uthark) [inline lib/tasks/maintenance.rake:52]: if we pass multiple with comma, how will it fork? IS delete_group would be an array then?

File: lib/tasks/maintenance.rake (hunk 1 above)
Comment (lara) [reply lib/tasks/maintenance.rake:47]: copy pasta from other task

File: lib/tasks/maintenance.rake (hunk 2 above)
Comment (lara) [reply lib/tasks/maintenance.rake:48]: yeah I can see how that's not a good idea `alias watch_the_world_burn="rake maintenance:delete_deploy_groups DELETE=pod17,pod18,pod19,pod20"` :D

File: lib/tasks/maintenance.rake (hunk 3 above)
Comment (lara) [reply lib/tasks/maintenance.rake:52]: it wouldn't have but we'll do one group at a time for safety anyways

**Review by grosser — APPROVED** [review]

**Pull request merged**